hypothesis-python/tests/cover/test_targeting.py

```diff
@@ -0,0 +1,91 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis/
+#
+# Most of this work is copyright (C) 2013-2019 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import absolute_import, division, print_function
+
+import pytest
+
+from hypothesis import example, given, strategies as st, target
+from hypothesis.errors import InvalidArgument
+from hypothesis.internal.compat import string_types
+
+
+@example(0.0, "this covers the branch where context.data is None")
+@given(observation=st.floats(allow_nan=False, allow_infinity=False), label=st.text())
+def test_allowed_inputs_to_target(observation, label):
+    target(observation, label)
+
+
+@given(
+    observation=st.floats(min_value=1, allow_nan=False, allow_infinity=False),
+    label=st.sampled_from(["a", "few", "labels"]),
+)
+def test_allowed_inputs_to_target_fewer_labels(observation, label):
+    target(observation, label)
+
+
+@given(st.floats(min_value=1, max_value=10))
+def test_target_without_label(observation):
+    target(observation)
+
+
+@given(
+    st.lists(
+        st.tuples(
+            st.floats(allow_nan=False, allow_infinity=False),
+            st.sampled_from(["a", "few", "labels"]) | st.text(),
+        ),
+        min_size=1,
+    )
+)
+def test_multiple_target_calls(args):
+    for observation, label in args:
+        target(observation, label)
+
+
+@given(
+    st.lists(st.floats(allow_nan=False, allow_infinity=False), min_size=5, max_size=10)
+)
+def test_respects_max_pool_size(observations):
+    """Using many examples of several labels like this stresses the
+    pool-size logic in TargetSelector, and its internal assertions.
+    """
+    for label, obs in enumerate(observations):
+        target(obs, str(label))
+
+
+def everything_except(type_):
```

Comment:
Nit: I'm not really wild about tests like this. I don't think they add much over example based tests and they're moderately expensive.

Reply:
I generally agree, but thought it was worth it in this case as `target()` explicitly has different behaviours in non-`@given` tests and between phases of an `@given` test.  So if we're using Hypothesis anyway, `everything_except` seemed the natural tactic.

I've added a comment that this is considered an unusual case though.